stgit/commands/refresh.py

```diff
@@ -354,10 +354,25 @@ def func(parser, options, args):
                 'To force a full refresh use --force.'
             )
 
-    # Commit index to temp patch, and absorb it into the target patch.
-    retval, temp_name = make_temp_patch(
-        stack, patch_name, paths, temp_index=path_limiting
-    )
+    # Update index and write tree
+    tree = write_tree(stack, paths, temp_index=path_limiting)
+
+    # Run pre-commit hook, if fails, abort refresh
+    if not options.no_verify:
+        pre_commit_success = run_pre_commit_hook(stack.repository)
+        if not pre_commit_success:
+            raise CmdException(
+                'pre-commit hook failed, review the changes using `stg diff`'
+                ', run `stg add` to add them to index and run `stg refresh` again'
+            )
+
+    # Update index and rewrite tree if pre-commit hook updated files in index
+    if not stack.repository.default_index.is_clean(tree):
```

Comment:
@jpgrayson I am not sure if this is correct. I checked almost all cases and I am getting correct result. But still I have feeling this might be a bug since I am comparing with the default index for checking if hook has changed something.

What do you think?

Reply:
I agree that this doesn't feel super robust because of the use of `default_index`. In practice, the default index would only possibly be incorrect when using path limiting, i.e. because that is the only case where a temporary index is used instead of the default index. However, like you, I was unable to craft a test where this code yielded clearly incorrect behavior.

I also spent some time experimenting with `git refresh` and how it's behavior with pre-commit hooks; with and without a pathspec; and with or without changes staged. The semantics of how `git commit` deals with a `pre-commit` script modifying its temp index when using pathspec are non-obvious and, frankly, a bit surprising. It is not clear to me that there is a guiding light from `git refresh` that `stg refresh` should or needs to follow.

In other words, I think this patch as-is is okay. It *might* be more correct to run the pre-commit script in the context of the temp index when path-limiting is in effect (i.e. by setting `GIT_INDEX_FILE`), but again, I do not have a test that shows clearly incorrect behavior when we do not set `GIT_INDEX_FILE` for `pre-commit`.

The one possible change to these two particular lines of code would be to only execute them if we have actually run the pre-commit hook. This would require unrolling `run_pre_commit_hook()` here:

```python
    # Run pre-commit hook, if fails, abort refresh
    if not options.no_verify:
        pre_commit_hook = get_hook(
            stack.repository,
            'pre-commit',
            extra_env={} if options.edit else {'GIT_EDITOR': ':'},
        )
        if pre_commit_hook:
            try:
                pre_commit_hook()
            except RunException:
                raise CmdException(
                    'pre-commit hook failed, review the changes using `stg diff`, '
                    'run `stg add` to add them to index and run `stg refresh` again'
                )
            else:
                # Update index and rewrite tree if hook updated files in index
                if not stack.repository.default_index.is_clean(tree):
                    tree = write_tree(stack, paths, temp_index=path_limiting)
```

And  again, thank you for the thorough work you've done on this feature. It's very much appreciated!

Reply:
Thanks Pete. It was a pleasure working on my first open source contribution. If you think there is something relevant for me to implement (features etc) I would be happy to take it up.